ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,577 @@
+     * been deleted.
+     */
+    GET_DELETED_COORD(false, true),
+    /**
+     * Wait for the operation chain's blob ID to be reported as created on all replicas. Continue with the remaining
+     * actions in the operation chain afterwards.
+     */
+    AWAIT_CREATION(false, false),
+    /**
+     * Wait for the operation chain's blob ID to be reported as deleted on all replicas. Continue with the remaining
+     * actions in the operation chain afterwards.
+     */
+    AWAIT_DELETION(false, false);
+
+    /**
+     * {@code true} if the non-blocking router should be used, {@code false} if the coordinator-backed router should
+     * be used.
+     */
+    final boolean nonBlockingRouter;
+    /**
+     * {@code true} if this operation follows a delete operation.
+     */
+    final boolean afterDelete;
+
+    OperationType(boolean nonBlockingRouter, boolean afterDelete) {
+      this.nonBlockingRouter = nonBlockingRouter;
+      this.afterDelete = afterDelete;
+    }
+  }
+
+  /**
+   * Describes an operation chain and provides useful metadata and operation results for testing.
+   */
+  static class OperationChain {
+    final int chainId;
+    final BlobProperties properties;
+    final byte[] userMetadata;
+    final byte[] data;
+    final Queue<OperationType> operations;
+    final List<TestFuture> testFutures = new ArrayList<>();
+    final CountDownLatch latch = new CountDownLatch(1);
+    String blobId;
+
+    OperationChain(int chainId, BlobProperties properties, byte[] userMetadata, byte[] data,
+        Queue<OperationType> operations) {
+      this.chainId = chainId;
+      this.properties = properties;
+      this.userMetadata = userMetadata;
+      this.data = data;
+      this.operations = operations;
+    }
+  }
+
+  /**
+   * This class encapsulates a future and allows the user to define a check method that runs tests on
+   * the retrieved value in the future.
+   *
+   * @param <T> The type of the encapsulated future
+   */
+  private static abstract class TestFuture<T> {
+    final Future<T> future;
+    final String operationType;
+    final OperationChain opChain;
+
+    TestFuture(Future<T> future, String operationType, OperationChain opChain) {
+      this.future = future;
+      this.operationType = operationType;
+      this.opChain = opChain;
+    }
+
+    /**
+     * Generate a name for the tested operation
+     * @return the operation name
+     */
+    String getOperationName() {
+      return operationType + "-" + opChain.chainId;
+    }
+
+    /**
+     * Return the value inside the future or throw an {@link AssertionError} if an exception
+     * occurred.
+     * @return the value inside the future
+     */
+    T get() {
+      try {
+        return this.future.get();
+      } catch (Exception e) {
+        e.printStackTrace();
+        Assert.fail("Exception occured in operation: " + getOperationName() + ", exception: " + e);
+        return null;
```

Comment:
The exception is now rethrown, so the compiler no longer requires a return value past this point.